third_party/amd/lib/TritonAMDGPUToLLVM/ElementwiseOpToLLVM.cpp

```diff
@@ -988,112 +834,80 @@ struct FpToFpOpConversion
-#else
-        {{F16TyID, F8E4M3B15TyID}, Fp16_to_Fp8E4M3B15(computeCapability >= 80)},
-        {{F16TyID, F8E4M3TyID}, Fp16_to_Fp8E4M3Nv},
-        {{F16TyID, F8E5M2TyID}, Fp16_to_Fp8E5M2(computeCapability >= 90)},
-#endif
-    // F8 -> BF16
-#ifdef USE_ROCM
-      {{F8E5M2TyID, BF16TyID}, Fp8E5M2_to_Bf16},
-#else
-        {{F8E5M2TyID, BF16TyID}, Fp8E5M2_to_Bf16(computeCapability >= 90)},
-        {{F8E4M3TyID, BF16TyID}, Fp8E4M3Nv_to_Bf16},
-#endif
-    // BF16 -> F8
-#ifdef USE_ROCM
-      {{BF16TyID, F8E5M2TyID}, Bf16_to_Fp8E5M2},
-#else
-        {{BF16TyID, F8E5M2TyID}, Bf16_to_Fp8E5M2(computeCapability >= 90)},
-        {{BF16TyID, F8E4M3TyID}, Bf16_to_Fp8E4M3Nv},
-        // F32 -> F8
-        {{F32TyID, F8E4M3TyID}, Fp32_to_Fp8E4M3Nv},
-        {{F32TyID, F8E5M2TyID}, Fp32_to_Fp8E5M2},
-#endif
-    };
-    std::pair<TypeID, TypeID> key = {srcTy.getTypeID(), dstTy.getTypeID()};
+
+    auto undefRounding = static_cast<RoundingMode>(-1);
+
+    static DenseMap<std::tuple<TypeID, TypeID, RoundingMode>, ConverterT>
+        srcMap = {
+            // F8 -> F16
+            {{F8E4M3FNUZTyID, F16TyID, undefRounding},
+             Fp8E4M3FNUZ_to_Fp16(computeCapability)},
+            {{F8E5M2FNUZTyID, F16TyID, undefRounding},
+             Fp8E5M2FNUZ_to_Fp16(computeCapability)},
+            {{F8E5M2TyID, F16TyID, undefRounding}, Fp8E5M2_to_Fp16},
+            // F16 -> F8
+            {{F16TyID, F8E5M2FNUZTyID, RoundingMode::RTNE},
+             Fp16_to_Fp8E5M2FNUZ(computeCapability)},
+            {{F16TyID, F8E4M3FNUZTyID, RoundingMode::RTNE},
+             Fp16_to_Fp8E4M3FNUZ(computeCapability)},
+            {{F16TyID, F8E5M2TyID, RoundingMode::RTNE}, Fp16_to_Fp8E5M2_RTNE},
+            {{F16TyID, F8E5M2TyID, RoundingMode::RTZ}, Fp16_to_Fp8E5M2_RTZ},
+            // F8 -> BF16
+            {{F8E5M2TyID, BF16TyID, undefRounding}, Fp8E5M2_to_Bf16},
+            {{F8E5M2FNUZTyID, BF16TyID, undefRounding}, Fp8E5M2FNUZ_to_Bf16},
+            {{F8E4M3FNUZTyID, BF16TyID, undefRounding}, Fp8E4M3FNUZ_to_Bf16},
+            // BF16 -> F8
+            {{BF16TyID, F8E5M2TyID, RoundingMode::RTNE}, Bf16_to_Fp8E5M2},
+            {{BF16TyID, F8E5M2FNUZTyID, RoundingMode::RTNE},
+             Bf16_to_Fp8E5M2FNUZ},
+            {{BF16TyID, F8E4M3FNUZTyID, RoundingMode::RTNE},
+             Bf16_to_Fp8E4M3FNUZ},
+
+            // F32 <-> F8
+            {{F32TyID, F8E4M3FNUZTyID, RoundingMode::RTNE},
+             Fp32_to_Fp8E4M3FNUZ},
+            {{F32TyID, F8E5M2FNUZTyID, RoundingMode::RTNE},
+             Fp32_to_Fp8E5M2FNUZ},
+            {{F8E4M3FNUZTyID, F32TyID, undefRounding}, Fp8E4M3FNUZ_to_Fp32},
+            {{F8E5M2FNUZTyID, F32TyID, undefRounding}, Fp8E5M2FNUZ_to_Fp32},
+        };
+    std::tuple<TypeID, TypeID, RoundingMode> key = {
+        srcTy.getTypeID(), dstTy.getTypeID(),
+        roundingMode.value_or(undefRounding)};
+    if (srcMap.count(key) == 0) {
+      llvm::errs() << "Unsupported conversion from " << srcTy << " to "
```

Comment:
We should not directly call into `llvm::errs()` which is stderrr pretty much. im mlir there are good facilities for error reporting. You can change this funciton to return `FailureOr<..>` and `return emitError(..)` here. In where this function is used, check the result `if (failed(...))`. Per my understanding if `createDestOps` return and empty `SmallVector` will cause the pass to fail.

Reply:
Do you mean changing this function to return `FailureOr<...>` and then check the result where this function is used. And if it fails, then we can do `emitError(...)` with the `loc` information.

Reply:
@antiagainst I did a refactoring but not sure if I was doing the `emitError` appropriately. Could you take a look at the latest commit and let me know if you see issues or concerns?